cmd/ndm_daemonset/probe/mountprobe.go

```diff
@@ -106,3 +148,85 @@ func (mp *mountProbe) FillBlockDeviceDetails(blockDevice *blockdevice.BlockDevic
 		blockDevice.FSInfo.FileSystem = basicMountInfo.FileSystem
 	}
 }
+
+func (mp *mountProbe) setupEpoll() error {
+	ep, err := epoll.New()
+	if err != nil {
+		return err
+	}
+	mp.epoll = &ep
+	return ep.AddWatcher(epoll.Watcher{
+		FileName:   mp.mountsFileName,
+		EventTypes: []epoll.EventType{epoll.EPOLLERR, epoll.EPOLLPRI},
+	})
+}
+
+func (mp *mountProbe) listen() {
+	eventCh, err := mp.epoll.Start()
+	if err != nil {
+		klog.Errorf("error while starting epoll: %v", err)
+		return
+	}
+	defer mp.epoll.Stop()
+	klog.Info("started mount change detection loop")
+	defaultMsg := controller.EventMessage{
+		Action:          string(MountEA),
+		Devices:         nil,
+		AllBlockDevices: true,
+	}
+
+	for range eventCh {
+		// regenerate mounts table and get the changes
+		newMountTable, err := mp.newMountTable()
+		if err != nil {
+			klog.Error("failed to generate mounts table.")
+			mp.destination <- defaultMsg
+		}
+		mtDiff := libmount.GenerateDiff(mp.mountTable, newMountTable)
+		mp.mountTable = newMountTable
+		mp.processDiff(mtDiff)
+	}
+}
+
+func (mp *mountProbe) newMountTable() (*libmount.MountTab, error) {
+	return libmount.NewMountTab(libmount.FromFile(mp.mountsFileName,
+		libmount.MNT_FMT_FSTAB),
+		libmount.WithAllowFilter(libmount.SourceContainsFilter("/dev/")),
+		libmount.WithDenyFilter(libmount.SourceFilter("overlay")),
+		libmount.WithDenyFilter(libmount.TargetContainsFilter("/var/lib/kubelet/pod")))
+}
+
+func (mp *mountProbe) processDiff(diff libmount.MountTabDiff) {
```

Comment:
Can this functionality also be moved to libmount package?

Reply:
Just the listing of devices in the libmount package and the rest of functionality related to sending messages can be here.

Reply:
I feel it should stay here. this is because the task of processing the changes and sending an appropriate event to the `UdevEventMessageChannel` should be the task of the probe package. `libmount` should be agnostic of this as a package according to me. Some other applications might use the difference for taking some other kind of action.

Reply:
since this is related to operations with the mount, I feel libmount should have it. Also the correct separation of functionality should be there,  this function should just process the diff and return the changed devices with the action. Then those set of devices can be passed on to the channel from the caller of the diff function.

Reply:
oh ok, you mean the processing which devices changed part. So that too should be a part of the mountprobe maybe, since the diff generated by `libmount` is complete in the sense that it tells what filesystems changed (device, mount-point, and the action). The processing of diff in mount probe is only done because we are not concerned about the mount-point component, and just want to extract the devices that changed. But having said that, I think we can have a utility method for `MountTabDiff` to give a list of changed devices.